### ISSUE TITLE
[R-305] Incorrect API key provided Azure OpenAI for TestsetGenerator

### ISSUE THREAD

**Issue by llvll0hsen** [explodinggradients/ragas#1526]
I am trying to use Azure Openai to create a test dataset. But for some reasons the `generate_with_langchain_docs` give the following error:

`unable to apply transformation: Error code: 401 - {'error': {'message': 'Incorrect API key provided: 0ba2ec70********************407d. You can find your API key at https://platform.openai.com/account/api-keys.', 'type': 'invalid_request_error', 'param': None, 'code': 'invalid_api_key'}}
`
when I test my api key using `azure_llm.invoke("tell me a jok about weather")` it work.

here is my code
  
```
import os
from ragas.testset import TestsetGenerator
from ragas.llms import LangchainLLMWrapper
from ragas.embeddings import LangchainEmbeddingsWrapper

from langchain_openai.chat_models import AzureChatOpenAI
from langchain_openai.embeddings import AzureOpenAIEmbeddings
from langchain_community.document_loaders import WebBaseLoader


config = {
    "base_url": "https://xxxxx.openai.azure.com",
    "model_deployment": "my-gpt-4o-rag",
    "model_name": "gpt-4o",
    "embedding_deployment": "my-text-embedding-ada-002",
    "embedding_name": "text-embedding-ada-002",  }

loader = WebBaseLoader("https://www.advancinganalytics.co.uk/")
documents = loader.load()
for document in documents:
    document.metadata['filename'] = document.metadata['source']

azure_llm = AzureChatOpenAI(
    openai_api_version= "2024-08-01-preview",
    azure_endpoint=config["base_url"],
    azure_deployment=config["model_deployment"],
    model=config["model_name"],
    validate_base_url=False,
)

response = azure_llm.invoke("tell me a jok about the weather") #---> this works

azure_llm = LangchainLLMWrapper(azure_llm)
generator = TestsetGenerator.from_langchain(azure_llm)
testset = generator.generate_with_langchain_docs(
    documents,
    testset_size=5,
)

```

`Ragas=0.2.1` and `langchain_openai==0.2.2`

<sub>[R-305](https://linear.app/exploding-gradients/issue/R-305/incorrect-api-key-provided-azure-openai-for-testsetgenerator)</sub>

**Comment by llvll0hsen** [comment]
I've already check the api key by calling the `AzureChatOpenAI.invoke` method and it works just fine.

**Comment by MariosKef** [comment]
@llvll0hsen I have exactly the same issue! It seems that the problem lies in the `llm_factory`.

**Comment by Kefan-pauline** [comment]
Same here.

**Comment by jjmachan** [comment]
hey let me look into this at this ASAP and I will get back

**Comment by Kefan-pauline** [comment]
> hey let me look into this at this ASAP and I will get back

Thanks! BTW, it is the same for embedding_factory. It would be nice to pass the llm and embedding model through the TestsetGenerator parameter and actually **use them** in the default transforms.

**Comment by hunter-walden2113** [comment]
I was having the same error. It seems like the llm_factory defaults to OpenAI and there's no support in that function for AzureOpenAI. I altered the code locally to try to see if it would accept an AzureOpenAI client but no luck. In the previous version (0.1.21) you used to be able to pass your LLM and embedding client directly to TestsetGenerator, but that seems to have changed in the latest version.

**Comment by MariosKef** [comment]
Are there any updates on this?

**Comment by hunter-walden2113** [comment]
I figured out the problem. I implemented a temporary fix locally just to have the ability to produce a test set, but it is not a long-term solution. I am willing to spend some time working on a fix if nobody else is.

**Comment by jjmachan** [comment]
hey @hunter-walden2113 really curious about what the fix was - I'll also take this up this week

**Comment by hunter-walden2113** [comment]
@jjmachan : the problem originates here: 

`ragas.testset.synthesizers.generate.TestsetGenerator.generate_with_langchain_docs`. 

The current code defaults to using `default_transforms()` if the user does not pass an argument for the `transforms` parameter. 
When default_transforms() is called, it instantiates all the required `Extractor` objects in transforms.__init__.py, where the default llm in the `Extractor` class is instantiated by calling `llm_factory`. Since the user is not using the dafault llm (OpenAI), and therefore not setting the required environmental variables, they are getting the "API Key not found error". 

I fixed this by allowing the user to pass their models into TestsetGenerator, and providing those model clients to the Extractors, instead of using the default llms. 

To fix this I did the following:

1. Added `embedding_model` as an attribute to `TestsetGenerator`.
2. Added `embedding_model: LangchainEmbeddings` as a parameter to `TestsetGenerator.from_langchain`
3. Changed the return from `TestsetGenerator.from_langchain` to `return cls(LangchainLLMWrapper(llm), LangchainEmbeddingsWrapper(embedding_model), knowledge_graph)`
4.  Added both an `llm` and `embedding_model` parameter to `TestsetGenerator.generate_with_langchain_docs`
5. Added `llm` and `embedding_model` as parameters to `testset.transforms.__init__.default_transforms`
6. Passed the llm and embedding arguments to each of the Extractors in default_transforms as appropriate. 

I will schedule some time to talk with you to see what you think of the fix and how to improve it.
